SceneGadget : Support alternative renderers

johnhaddon opened the pull request:
The SceneGadget is the primary component in the 3D viewer, responsible for drawing the scene in OpenGL. This adds a `setRenderer()` method to it, providing the option of doing a hybrid render where OpenGL is used for drawing bounds and visualisers and, say, Arnold is used for rendering everything else. Because everything is hidden behind the SceneGadget API, existing tools for selection, snapping etc all continue to work as before.

This PR doesn't expose the feature to users yet - that will come in a followup PR which adds appropriate render settings to the SceneView. But testing is provided by the new ArnoldSceneGadgetTest, which runs all the original OpenGL SceneGadget tests in the new hybrid mode.

danieldresser-ie commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -290,6 +429,55 @@ bool SceneGadget::objectAt( const IECore::LineSegment3f &lineInGadgetSpace, Gaff
+	return true;
+}
+
+bool SceneGadget::openGLObjectAt( const IECore::LineSegment3f &lineInGadgetSpace, GafferScene::ScenePlug::ScenePath &path, float &depth ) const
```
Logic seems sound, but I find the names a little misleading:  openGLObjectAt and objectAt has such similar names that it feels weird they don't have the same return values.  Because openGLObjectAt is private, it doesn't really matter.

danieldresser-ie commented on src/GafferSceneUI/OutputBuffer.cpp:
```diff
@@ -0,0 +1,493 @@
+	outColor = mix(
+		outColor,
+		vec4( 0.466, 0.612, 0.741, 1.0 ) * outColor.a,
+		0.75 * float( contains( selectionTexture, id ) )
```
I would have expected this to be done as a preprocess or something, rather than doing a binary search per fragment, but if it works, it works.  Have you tested if the slowdown gets noticeable when selecting very large numbers of objects?  Oh, or if you're only using 1 dimension of the texture, I guess you would just run out of selection indices if you tried to select half a particle cloud?

johnhaddon replied:
Used a buffer texture in https://github.com/GafferHQ/gaffer/pull/4674/commits/b79b4eb073a9a2957e605f0b8a81060b3c6acf0c.

johnhaddon replied:
Also opened https://github.com/ImageEngine/cortex/pull/1264 that will remove the need for the texture unit workaround in that commit. I'd like to get this merged without having to do a Cortex release though - will remove the workaround later.

danieldresser-ie reviewed (commented):
Not guaranteeing I've been completely thorough, but I've tried to give everything a basic once over to keep things moving, and it looks like a good foundation.  The internal details can always be iterated on later if we need to.

johnhaddon commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -123,6 +123,56 @@ Box3f sceneBound( const ScenePlug *scene, const PathMatcher *include, const Path
 	);
 }
 
+vector<IECore::TypeId> typeNamesForTypeIds( const vector<string> &typeNames )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/4674/commits/4a16223a05f438ed523c67ef138a4243441f7da3

danieldresser-ie commented:
Looks good, aside from one naming thing:  

Now that you've named it "From" instead of "For", I'm feeling more confident that the name `typeNamesFromTypeIds` is backwards? It takes names and returns ids, by calling idFromName on each element?

johnhaddon commented:
> Now that you've named it "From" instead of "For", I'm feeling more confident that the name typeNamesFromTypeIds is backwards?

Sorry, I was being a complete numpty - I should have understood from your first description. Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/4674/commits/9d33966d2f9a90f3b3d6e87b1aaad68b2f47810f.
